### PR TITLE
Support recording angular momentum

### LINKED ISSUE #197: Record angular momentum
### Tasks
- [ ] Record total angular momentum in `Record__Conservation`
- [ ] User-specified center, which can change in time
- [ ] Support `HYDRO, ELBDM, PARTICLE`

### PR DESCRIPTION
**Pull request by hsinhaoHHuang**
### Goal
- To resolve #197 

### Changes
- Record angular momentum (`AngMomX/Y/Z_Gas`, `AngMomX/Y/Z_Par`, `AngMomX_All`) in `Record__Conservation` when `OPT__CK_CONSERVATION` is on.

- Add the runtime option `ANGMOM_ORIGIN_X/Y/Z` to specify the origin coordinates to compute the angular momentum. 
   - The default is the box center.
   - The center cannot change in time so far.

- Record center of mass (`CoMX/Y/Z_Gas`, `CoMX/Y/Z_Par`, `CoMX/Y/Z_All`) of the whole box in `Record__Conservation` when `OPT__CK_CONSERVATION` is on.
   - Since the relative error of the center of mass coordinates doesn't make much sense, the average velocity of the center of mass is output instead.

### Validation
- A Python script using `yt` to compute the angular momentum (and other conserved quantities) is added in `example/test_problem/Hydro/Plummer/plot_script/yt_get_conserved_quantities.py`.
   - The results are consistent with `Record__Conservation` for the `Plummer` test problem.
- The previous Python script using `yt` to find the center of mass `example/test_problem/Hydro/Plummer/plot_script/yt_find_center.py` has been updated to find the center of mass of `gas`, `particle`, and `all` separately.
   - The results are consistent with `Record__Conservation` for the `Plummer` test problem.

### CODE REVIEW

**Comment by hsinhaoHHuang** [comment]
Implementing recording the center of mass in `Record__Conservation` has been done.

File: example/test_problem/Hydro/Plummer/plot_script/yt_find_center.py [hunk 1]
```diff
@@ -30,72 +30,82 @@
...
 
 
-field_dens       = ('gamer','Dens')
-field_pardens    = ('gamer','ParDens')
-field_totdens    = ('gamer','TotDens')
-field_pote       = ('gamer','Pote')
+field_dens       = ('gamer',    'Dens')
+field_pardens    = ('gamer', 'ParDens')
+field_totdens    = ('gamer', 'TotDens')
+field_pote       = ('gamer',    'Pote')
```
Comment (ChunYen-Chen) [inline example/test_problem/Hydro/Plummer/plot_script/yt_find_center.py:41]: Please notify the user if the `OPT__OUTPUT_POT` or not.

File: example/test_problem/Hydro/Plummer/plot_script/yt_get_conserved_quantities.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+didx        = args.didx
+prefix      = args.prefix
+
+
+# define new fields
+def _kinetic_energy(field, data):
+    return data['kinetic_energy_density']*data['cell_volume']
+
+def _potential_energy(field, data):
+    return 0.5*data['Pote']*data['Dens']*data['cell_volume']
```
Comment (ChunYen-Chen) [inline example/test_problem/Hydro/Plummer/plot_script/yt_get_conserved_quantities.py:37]: Please notify the user if the `OPT__OUTPUT_POT` or not.

File: src/Particle/Par_Aux_GetConservedQuantity.cpp [hunk 3]
```diff
@@ -24,49 +24,76 @@
...
+         Mass_ThisRank    += amr->Par->Mass[p];
+         CoMX_ThisRank    += amr->Par->Mass[p]*amr->Par->PosX[p];
+         CoMY_ThisRank    += amr->Par->Mass[p]*amr->Par->PosY[p];
+         CoMZ_ThisRank    += amr->Par->Mass[p]*amr->Par->PosZ[p];
+         MomX_ThisRank    += amr->Par->Mass[p]*amr->Par->VelX[p];
+         MomY_ThisRank    += amr->Par->Mass[p]*amr->Par->VelY[p];
+         MomZ_ThisRank    += amr->Par->Mass[p]*amr->Par->VelZ[p];
+         const double dX   = amr->Par->PosX[p] - ANGMOM_ORIGIN_X;
+         const double dY   = amr->Par->PosY[p] - ANGMOM_ORIGIN_Y;
+         const double dZ   = amr->Par->PosZ[p] - ANGMOM_ORIGIN_Z;
```
Comment (ChunYen-Chen) [inline src/Particle/Par_Aux_GetConservedQuantity.cpp:64]: Can you move this part of the code to line 55?

**Review by ChunYen-Chen — CHANGES_REQUESTED** [review]
The function works well! I only have some minor comments.

**Review by ChunYen-Chen — APPROVED** [review]

**Comment by hsinhaoHHuang** [comment]
@hyschive 
When we set `OPT__CORR_AFTER_ALL_SYNC  = 1 (every step)` in `Input__Parameter`,
the total momenta of the particles in `Record__Conservation` are the same as the results computed from `yt` for the final dump.

**Review by hyschive — COMMENTED** [review]
@hsinhaoHHuang Looks great! I only have minor comments. Once it's merged, please file a follow-up PR to merge `main` into `psidm` and to support angular momentum and CoM for ELBDM.

**Comment by hsinhaoHHuang** [comment]
@hyschive  Thanks for the review. I have made the updates to this pull request according to your suggestions.

**Review by hyschive — APPROVED** [review]

**Pull request merged**